File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py
# 아이디어 : dfs를 활용하여 경우의 수 구하기
# 알고리즘 : dfs

def solution(numbers, target):
    answer = 0
    n = len(numbers)
    
    def dfs(depth, num):
        nonlocal answer
        
        # 종료 조건
        if depth == n :
            if num == target :
                answer += 1
            return
        
        # 더하기 빼기 정의
        dfs(depth + 1, num + numbers[depth])
        dfs(depth + 1, num - numbers[depth])
        
    # 함수 실행
    dfs(0, 0)
    
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    
    int answer = 0;
    
    public int solution(int[] numbers, int target) {
        dfs(0, numbers, target, 0);
        return answer;
    }
    
    private void dfs(int depth, int[] numbers, int target, int sumValue) {
        // 종료조건 정의
        if (depth == numbers.length) {
            if (sumValue == target) {
                answer += 1;
            }
            return;
        }
        
        dfs(depth + 1, numbers, target, sumValue + numbers[depth]);
        dfs(depth + 1, numbers, target, sumValue - numbers[depth]);
    }
}